src/hash.c

```diff
@@ -0,0 +1,226 @@
+#include <pthread.h>
+
+#include "data.table.h"
+
+struct hash_pair {
+    SEXP key;
+    R_xlen_t value;
```

Comment:
Can halve the memory requirements by storing `int` indices into an array instead of `SEXP` pointers and `int` mark values instead of `R_xlen_t` (thanks to Sebastian Krantz for the idea). Cost: this makes the hash table incompatible with long vectors. Since `int` dimensions for matrices and `data.frame`s are baked into almost all of R, it's probably worth using `int`s here too.